Fix flaky test.

icemac opened the pull request:
~~Wait longer to hopefully circumvent the problem.~~

This was not a flaky test problem but a problem with `waitress >= 2` and `zope.interface` not having a final 5.3.0 release.

Commit:
Fix flaky test.

Wait longer to hopefully circumvent the problem.

icemac commented:
Tests fail locally, too and I see 
`pkg_resources.ContextualVersionConflict: (zope.interface 5.2.0 (/Users/mac/vcs/gocept/zope.paste/.tox/py39/lib/python3.9/site-packages), Requirement.parse('zope.interface>=5.3.0a1'), {'zope.component'})`

mgedmin commented on src/zope/paste/tests.py:
```diff
@@ -128,7 +128,7 @@ def test_serving_test_app():
     >>> server.start()
-    >>> import time; time.sleep(5)  # doctest: +ELLIPSIS
+    >>> import time; time.sleep(10)  # doctest: +ELLIPSIS
     Serving on http://localhost:...
```
What if the thread is faster than expected and prints the `Serving on ...` right after `server.start()`?

Ideally we'd have a synchronization primitive of some kind and could do `>>> server.start(); wait_for_server_to_be_ready()`.

mgedmin replied:
Ok, that's not what happens.  I'm not sure what happens -- waitress stopped printing 'Serving on ...' by default?

icemac replied:
It seems like `waitress >= 2` needs a custom logging setup to get it printed here.

icemac replied:
See https://github.com/zopefoundation/zope.paste/pull/3/commits/967f784c5aa40fa1fed3eea07a203fd829a84e5e.

Commit:
Fix tests.

`waitress >= 2` seems to need some extra hand holding to log the line `serving on ...`.

mgedmin commented on src/zope/paste/tests.py:
```diff
@@ -124,14 +124,27 @@ def test_basic_serve():
     >>> import threading
     >>> from zope.paste import serve
-    >>> server = threading.Thread(target=serve.serve, args=(['app.ini'],))
+    >>> app_ini = os.path.join(os.getcwd(), 'app.ini')
```
Isn't this better expressed as 
```suggestion
    >>> app_ini = os.path.abspath('app.ini')
```
?

icemac replied:
Let's try it.

mgedmin commented on src/zope/paste/tests.py:
```diff
@@ -124,14 +124,27 @@ def test_basic_serve():
+    Serving on http://...
 
-    >>> print(urlopen('http://localhost:{}/'.format(PORT)).read().decode())
+    >>> print(urlopen('http://127.0.0.1:{}/'.format(PORT)).read().decode())
```
Oh, that is interesting.

I once had problems where connections to localhost:NNNN failed but 127.0.0.1:NNNN succeeded, because my /etc/hosts had localhost resolve to ::1, but the service on port NNNN was listening on IPv4 only.

icemac replied:
When using `localhost` `waitress >= 2` opens both on IPv4 and IPv6 but this is not backwards compatible with `waitress 1.x` which is still needed for Python 2.7 as `waitress >= 2` dropped Python 2 support.

mgedmin commented on src/zope/paste/tests.py:
```diff
@@ -124,14 +124,27 @@ def test_basic_serve():
+    >>> import logging
+    >>> logger = logging.getLogger('waitress')
+    >>> logger.addHandler(logging.StreamHandler(sys.stdout))
+    >>> logger.setLevel(logging.INFO)
```
I'd rather do that in `setUp()`.

(And remove the handler in `tearDown()`, if we ever have two tests relying on waitress in some way, it won't be fun to have global state mutations persist from one test to another.)

icemac replied:
Good idea.

icemac replied:
> Good idea.

It seemed good, but it is too early. `doctest` changes `sys.stdout` to an internal stream. But this seems to be done after `setUp` is called. At least: it does not work when setting login up this way in `setUp`. So I added a teardown to the doctest, see https://github.com/zopefoundation/zope.paste/pull/3/commits/3d1e0ea9faa3ce8744e688ea1fbd2969edb0fe34.

mgedmin reviewed (commented):
Mostly LGTM

Commit:
Update src/zope/paste/tests.py

Co-authored-by: Marius Gedminas <marius@gedmin.as>

Commit:
Remove logging handler afterwards.

mgedmin reviewed (approved):
:+1:

icemac commented:
Thank you for reviewing this PR. 😃
